lib/mapObjects/CGTownInstance.cpp

```diff
@@ -1023,6 +994,55 @@ bool CGTownInstance::hasBuilt(const BuildingID & buildingID, FactionID townID) c
 	return false;
 }
 
+void CGTownInstance::addBuilding(const BuildingID & buildingID)
+{
+	if(buildingID == BuildingID::NONE)
+		return;
+
+	builtBuildings.insert(buildingID);
+	marketBuildingModeMapper(buildingID, [this](const EMarketMode mode) {addMarketMode(mode);});
+}
+
+void CGTownInstance::removeBuilding(const BuildingID & buildingID)
+{
+	if(!vstd::contains(builtBuildings, buildingID))
+		return;
+
+	builtBuildings.erase(buildingID);
+	marketBuildingModeMapper(buildingID, [this](const EMarketMode mode) {removeMarketMode(mode);});
+}
+
+void CGTownInstance::marketBuildingModeMapper(const BuildingID & buildingID, const std::function<void(const EMarketMode)> & func)
+{
+	const auto townType = (*VLC->townh)[getFaction()]->town;
+	if(townType->buildings.find(buildingID) == townType->buildings.end())
+		return;
+
+	// TODO how to remove hardcoded buildings?
```

Comment:
1. Add field `marketModes` to `CBuilding` class
2. Load this value from Json in `CTownHandler::loadBuilding`. See `MarketInstanceConstructor::initTypeData` for similar code.
3. Check for this field instead of checking for ID's in here
4. Update town building docs

Might be better to wait for my PR #4462 first though. I edited a lot of town buildings definitions & moved town building docs into a separate file

Reply:
Should the market repeat the same structure
`"types" : { "object" : { "modes": [] } }`  as [in](https://github.com/vcmi/vcmi/blob/develop/docs/modders/Map_Objects/Market.md) ?

Reply:
No need. Only inner-most "modes". Although would be better to name it "marketModes" or "tradeModes" or something similar since town building might have other functionality than market.

So something like this will do:
```json
"marketplace" : {
    "tradeModes" : [ "resource-resource" ]
}
```